### ISSUE TITLE
Type parameter leaks

### ISSUE THREAD

**Issue by andy-ms** [Microsoft/TypeScript#19854]
**TypeScript Version:**  2.7.0-dev.20171108

**Code**

```ts
declare function first<T>(array: T[]): T | undefined;
function f(x?: string[]) {
	return first(x);
}
const x = f([]);
```

**Expected behavior:**

Type of `x` is `{} | undefined`.

**Actual behavior:**

Type of `x` is `T | undefined`.

**Comment by mhegazy** [comment]
There is an error resolving the call for `first`. `x` of type `string[] | undefined` is not assignable to `T[]`. this goes through the resolveErrorCall..

**Comment by andy-ms** [comment]
Shouldn't we fill in type parameters in `resolveErrorCall`?

**Comment by mhegazy** [comment]
it is an error.. what difference does it make really..

**Comment by andy-ms** [comment]
Due to return type inference, the erroneous type can occur very far from the erroneous call site. A user won't know where to look first -- they'll just see `T` come out of nowhere, and since this is a common name they might think it refers to a different `T`.

**Comment by gcnew** [comment]
Considering `@ts-ignore`, this might lead to surprising results. Checking of the leaked type parameter is off.

```ts
declare function first<T>(array: T[]): T;
function mkSkolem(x?: string[]) {
    // @ts-ignore
    return undefined! && first(x);
}

const skolemWitness = mkSkolem(undefined);
const val: typeof skolemWitness = 5; // should be an error
```

**Comment by mhegazy** [comment]
`@ts-ignore` leads to surprising results, and should not be used IMHO.

**Comment by gcnew** [comment]
Sure, `@ts-ignore` is obviously a red mark. I'm concerned with the leaked parameter, though. It seems to behave like an implicit `any` that's not being flagged. If it were a type not assignable to/from anything, the leak would have been less bad.

**Comment by mhegazy** [comment]
>  that's not being flagged

It is.. there is an error!

**Comment by andy-ms** [comment]
#19620 will fix this. This was only a problem in services because we forgot to always fill in all the type arguments in the error case `pickLongestCandidateSignature` that only runs if we're in services.

**Comment by andy-ms** [comment]
Fixed by #25100
